### PR TITLE
Hj/save sub image

### PR DESCRIPTION
**Pull request by zarda**
Support the feature of exporting subimage: (co-work with backend [implement](https://github.com/CARTAvis/carta-backend/issues/618)) in #1175.

Add functions:
- Export original image
- Export subimage in a spectral range of original image
- Export subimage in a combo of stokes of original image
- Export a region of chopped original image
- Export a shared region of chopped original image
- Drop degenerated axes

Add interface:
File -> Save Image

[Reference](https://docs.google.com/document/d/1Jpju5UUzyFD98-iTgXomtmmHlfbWSHHokEVPCiA2t3I/edit?pli=1#)

(Update Demo Clip)

https://user-images.githubusercontent.com/10767557/110899891-d634e800-833c-11eb-92ba-17bbc8ff1e25.mov

https://user-images.githubusercontent.com/10767557/110899900-d9c86f00-833c-11eb-81b3-8cee1170698c.mov

### CODE REVIEW

File: src/components/FileInfo/FileInfoComponent.tsx [hunk 1]
```diff
@@ -1,15 +1,18 @@
...
+import { CARTA } from "carta-protobuf";
+import { TableComponent, TableComponentProps } from "components/Shared";
 import "./FileInfoComponent.scss";
+import { AppStore, FileBrowserStore } from "stores";
+import { SpectralSystem, SpectralType, SpectralUnit } from "models";
 
 export enum FileInfoType {
     IMAGE_FILE = "image-file",
     IMAGE_HEADER = "image-header",
+    CHOP_IMAGE = "chop-image",
```
Comment (veggiesaurus) [inline src/components/FileInfo/FileInfoComponent.tsx]: I don't think this really belongs in the file info component, since it's more about settings for saving a file. I suggest we move this into a separate component. I also think we should use `SAVE_SUBIMAGE` instead of `CHOP_IMAGE`

File: src/services/BackendService.ts [hunk 2]
```diff
@@ -1,9 +1,14 @@
-import {action, observable, makeObservable, runInAction} from "mobx";
-import {CARTA} from "carta-protobuf";
-import {Observable, Observer, Subject, throwError} from "rxjs";
-import {AppStore, PreferenceStore, RegionStore} from "stores";
-import {ApiService} from "./ApiService";
-import {mapToObject} from "utilities";
+import { action, observable, makeObservable, runInAction } from "mobx";
+
```
Comment (veggiesaurus) [inline src/services/BackendService.ts]: where do these new lines come from?

File: src/stores/AppStore.ts [hunk 3]
```diff
@@ -1,11 +1,29 @@
...
+
 import * as Long from "long";
-import {Classes, Colors, IOptionProps, setHotkeysDialogProps} from "@blueprintjs/core";
-import {Utils} from "@blueprintjs/table";
+
 import * as AST from "ast_wrapper";
+
 import * as CARTACompute from "carta_computation";
-import {CARTA} from "carta-protobuf";
+
```
Comment (veggiesaurus) [inline src/stores/AppStore.ts]: whitespace changes

File: src/stores/FileBrowserStore.ts [hunk 4]
```diff
@@ -367,24 +423,40 @@ export class FileBrowserStore {
...
     }
 
-    constructor() {
-        makeObservable(this);
-        this.exportCoordinateType = CARTA.CoordinateType.WORLD;
-        this.exportFileType = CARTA.FileType.CRTF;
+    /// Transfer the request from stokes string
+    /// to [start, end, stride]
+    @computed get saveStokesRange(): number[] {
+        // [start, end, stride] for "ABCD"
```
Comment (veggiesaurus) [inline src/stores/FileBrowserStore.ts:473]: if it's stokes, we should use IQUV instead of ABCD

File: src/stores/FileBrowserStore.ts [hunk 5]
```diff
@@ -78,6 +105,9 @@ export class FileBrowserStore {
         if (AppStore.Instance.activeFrame && mode === BrowserMode.SaveFile) {
             this.saveFilename = AppStore.Instance.activeFrame.frameInfo.fileInfo.name;
         }
+        this.updateIniSaveSpectralRange();
```
Comment (veggiesaurus) [inline src/stores/FileBrowserStore.ts]: could we not use a `@computed` value (`automaticSpectralRange` instead of this and an autorun?)
I don't like `IniSaveSpectralRange`. If `Ini` is short for `Initial`, rather use `Init` (or just use `Initial`)

File: src/stores/FileBrowserStore.ts [hunk 6]
```diff
@@ -62,8 +69,28 @@ export class FileBrowserStore {
     @observable catalogFileInfo: CARTA.ICatalogFileInfo;
     @observable catalogHeaders: Array<CARTA.ICatalogHeader>;
 
+    // Save image
     @observable saveFilename: string = "";
     @observable saveFileType: CARTA.FileType = CARTA.FileType.CASA;
+    @observable saveSpectralRange: number[] = [0, 0];
+    @observable saveStokesOption: number;
+    @observable saveRegionId: number;
+    @observable isDropDegeneratedAxes: boolean;
```
Comment (veggiesaurus) [inline src/stores/FileBrowserStore.ts]: rename to `shouldDropDegenerateAxes`

File: src/services/BackendService.ts (hunk 2 above)
Comment (zarda) [reply src/services/BackendService.ts]: It comes from TSlint. Do we need this lint?

File: src/stores/FileBrowserStore.ts (hunk 4 above)
Comment (zarda) [reply src/stores/FileBrowserStore.ts:473]: It's only a semantic demo to quick explain the orders. Will use "IQUV" insteadly.

File: src/stores/FileBrowserStore.ts (hunk 6 above)
Comment (zarda) [reply src/stores/FileBrowserStore.ts]: Got it.

File: src/stores/FileBrowserStore.ts (hunk 5 above)
Comment (zarda) [reply src/stores/FileBrowserStore.ts]: There are several situations we need to initialize SaveSpectralRange: activeFrame changes, updateSpectralSystem, updateSpectralCoordinate, and start a new FileBrowserDialog. Suppose to use @reaction and @action to trigger the update of SaveSpectralRange.

Comment (zarda) [reply src/stores/FileBrowserStore.ts]: Changed the naming

File: src/components/FileInfo/FileInfoComponent.tsx (hunk 1 above)
Comment (zarda) [reply src/components/FileInfo/FileInfoComponent.tsx]: Changed the naming

Comment (zarda) [reply src/components/FileInfo/FileInfoComponent.tsx]: How about add a new component for image save, 'ImageSaveComponent' or 'FileSaveComponent'?

**Comment by kswang1029** [comment]
Cannot edit value (eg, -20). The value persists at 0.

https://user-images.githubusercontent.com/20819712/108675147-5841a600-7521-11eb-85a2-f7bc8f2b22c4.mp4

**Comment by kswang1029** [comment]
The value validation is triggered annoyingly too fast. 

https://user-images.githubusercontent.com/20819712/108676203-d3f02280-7522-11eb-86c9-66ba73a5c5ce.mp4

**Comment by zarda** [comment]
> The value validation is triggered annoyingly too fast.
> 
>  Feb-22-2021.15-29-36.mp4

How about increasing debounce timeout to 2000 ms?

File: src/services/BackendService.ts (hunk 2 above)
Comment (veggiesaurus) [reply src/services/BackendService.ts]: which tslint rule in particular? I don't mean the spaces between `{` and `action`, those are fine. I mean the blank lines between imports.

Comment (zarda) [reply src/services/BackendService.ts]: get the problem and correct it.

File: src/stores/AppStore.ts (hunk 3 above)
Comment (zarda) [reply src/stores/AppStore.ts]: Correct it

File: src/components/FileInfo/FileInfoComponent.tsx (hunk 1 above)
Comment (zarda) [reply src/components/FileInfo/FileInfoComponent.tsx]: May I have any hint about how to refactor it or any suggestion? I suppose to move out the codes related to 'save image' to a new component, 'components/SaveFile/ImageSaveComponent'.

Comment (veggiesaurus) [reply src/components/FileInfo/FileInfoComponent.tsx]: I'd say `FileBrowser/ImageSaveComponent` would work

Comment (zarda) [reply src/components/FileInfo/FileInfoComponent.tsx]: Got it.

**Comment by kswang1029** [comment]
Value validation is still too aggressive as it modified the other field instead of the current field only. Maybe it is simpler to just validate values at once after clicking the Save button and use a popup to highlight errors.

https://user-images.githubusercontent.com/20819712/111099881-6b282300-8581-11eb-9f04-57e6f32010c2.mp4

**Comment by kswang1029** [comment]
I just realized that the current layout may be misleading so that users think the output spectral definition is what have been chosen from the GUI. Could you change the UI to the following? That should be less ambiguous. 
Screen Shot 2021-03-15 at 11 34 43 AM

**Comment by zarda** [comment]
@kswang1029 How about let the input box be in red while the input value is invalid? This means to change the `intent` of input box to 'danger' while it's invalid.
Screen Shot 2021-03-15 at 12 15 43

**Comment by kswang1029** [comment]
> @kswang1029 How about let the input box be in red while the input value is invalid? This means to change the `intent` of input box to 'danger' while it's invalid.
> Screen Shot 2021-03-15 at 12 15 43

that might work.

**Comment by kswang1029** [comment]
As saving an image may take some time (esp. for large cubes), we should disable the "Save" button once it is clicked for the first time, otherwise we can trigger multiple save requests.

**Comment by kswang1029** [comment]
It appears that if we firstly save a Stokes IV cube from an IQUV cube, then we cannot save the IV cube to I or V only.

**Comment by zarda** [comment]
> It appears that if we firstly save a Stokes IV cube from an IQUV cube, then we cannot save the IV cube to I or V only.

Could you have more details about how to reproduce this?

**Comment by kswang1029** [comment]
> > It appears that if we firstly save a Stokes IV cube from an IQUV cube, then we cannot save the IV cube to I or V only.
> 
> Could you have more details about how to reproduce this?

Try the steps such as:
1) load an cube with IQUV
2) save IV as image_IV.fits
3) load or append image_IV.fits
4) try to save as a new image from image_IV.fits and the Stokes dropdown seems to have IV only. There should be IV, I and V three options.

**Comment by zarda** [comment]
> > > It appears that if we firstly save a Stokes IV cube from an IQUV cube, then we cannot save the IV cube to I or V only.
> > 
> > 
> > Could you have more details about how to reproduce this?
> 
> Try the steps such as:
> 
> 1. load an cube with IQUV
> 2. save IV as image_IV.fits
> 3. load or append image_IV.fits
> 4. try to save as a new image from image_IV.fits and the Stokes dropdown seems to have IV only. There should be IV, I and V three options.

Would you mind checking the header to see if there are still 2 stokes?
In my testing, there are "IV", "I", and "V" in the options if the source image has two stokes.

**Comment by kswang1029** [comment]
This is a test IV image generated from IQUV cube. Once loaded in CARTA we can see the Stokes parameter is correctly displayed in the cursor info bar. The header is defined properly. But when we try to save this image as I or V only, the dropdown displays IV only.
[test.fits.zip](https://github.com/CARTAvis/carta-frontend/files/6139414/test.fits.zip)

**Comment by kswang1029** [comment]
in some cases, users might make the window shorter. Whee this happens, there is no way to access those UI elements not in the view. Should change the layout to be scrollable when needed. 

Screen Shot 2021-03-15 at 3 23 19 PM


In addition, I suggest to make the UI element "From-To" more responsive when width changes.

**Comment by kswang1029** [comment]
When the width is not enough, we cannot access the full UI. Can the UI elements be responsive by reducing the width first then if the width limit is hit, changing into two lines.
Screen Shot 2021-03-25 at 10 37 02 AM

**Comment by kswang1029** [comment]
The drop degenerate axes toggle should be only triggered by clicking either the text or the toggle itself. Now it can be triggered by clicking an empty area in the same row.

https://user-images.githubusercontent.com/20819712/112412005-c368dd00-8d58-11eb-8190-72c31d4e9869.mp4

**Comment by kswang1029** [comment]
The save button should be disabled if red box presents. 

Screen Shot 2021-03-25 at 10 58 48 AM

**Comment by kswang1029** [comment]
Due to the new save image tab, we need more space for the HDU dropdown.
https://user-images.githubusercontent.com/20819712/112413131-a208f080-8d5a-11eb-8ff9-a709e31610a1.mp4

**Comment by kswang1029** [comment]
> Due to the new save image tab, we need more space for the HDU dropdown.
> https://user-images.githubusercontent.com/20819712/112413131-a208f080-8d5a-11eb-8ff9-a709e31610a1.mp4

is this fixed?

**Comment by kswang1029** [comment]
> > Due to the new save image tab, we need more space for the HDU dropdown.
> > https://user-images.githubusercontent.com/20819712/112413131-a208f080-8d5a-11eb-8ff9-a709e31610a1.mp4
> 
> is this fixed?

using relative width breaks the layout when selecting a multi-HDU fits.
Screen Shot 2021-03-31 at 3 00 07 PM

Maybe you can try a fixed width other than 200px.

**Comment by zarda** [comment]
Use unit em to base on the font-size in global preference. It maybe better while changing size of box.

**Review by kswang1029 — APPROVED** [review]
Feature-wise this PR is ready👍

**Comment by veggiesaurus** [comment]
> Use unit em to base on the font-size in global preference. It maybe better while changing size of box.

Which global preference are you referring to?

**Comment by zarda** [comment]
> > Use unit em to base on the font-size in global preference. It maybe better while changing size of box.
> 
> Which global preference are you referring to?

Sorry,
I misunderstand the effect range of ‘rem’ and ‘em’. Which one would be better in this case?

**Comment by veggiesaurus** [comment]
> > > Use unit em to base on the font-size in global preference. It maybe better while changing size of box.
> > 
> > 
> > Which global preference are you referring to?
> 
> Sorry,
> I misunderstand the effect range of ‘rem’ and ‘em’. Which one would be better in this case?

I think it's fine to use `pix`, since we don't have any global font size. 

@zarda please fix the unnecessary whitespace changes mentioned above, and the merge conflicts. We should then be able to merge this in

File: src/components/FileInfo/FileInfoComponent.tsx [hunk 7]
```diff
@@ -4,13 +4,15 @@ import {action, makeObservable, observable} from "mobx";
 import {Button, ButtonGroup, ControlGroup, Divider, FormGroup, HTMLSelect, InputGroup, IOptionProps, NonIdealState, Popover, PopoverInteractionKind, PopperModifiers, Position, Pre, Spinner, Tab, TabId, Tabs, Text} from "@blueprintjs/core";
 import {FixedSizeList as List} from "react-window";
 import AutoSizer from "react-virtualized-auto-sizer";
-import {CARTA} from "carta-protobuf";
+import {CARTA } from "carta-protobuf";
```
Comment (veggiesaurus) [inline src/components/FileInfo/FileInfoComponent.tsx]: whitespace

File: src/components/FileInfo/FileInfoComponent.tsx [hunk 8]
```diff
@@ -19,7 +21,7 @@ export enum FileInfoType {
 @observer
 export class FileInfoComponent extends React.Component<{
     infoTypes: FileInfoType[],
-    HDUOptions?: {HDUList: IOptionProps[], handleSelectedHDUChange: (hdu: string) => void;},
+    HDUOptions?: { HDUList: IOptionProps[], handleSelectedHDUChange: (hdu: string) => void; },
```
Comment (veggiesaurus) [inline src/components/FileInfo/FileInfoComponent.tsx]: unnecessary whitespace change. We can change them in a separate issue (e.g. reformatting with prettier)

File: src/components/FileInfo/FileInfoComponent.tsx [hunk 9]
```diff
@@ -34,13 +36,13 @@ export class FileInfoComponent extends React.Component<{
     @observable searchString: string = "";
     @observable matchedIter: number = 0;
     @observable isMouseEntered: boolean = false;
-    
+
     private isSearchOpened: boolean = false;
     private matchedTotal: number = 0;
-    private matchedIterLocation: {line: number, num: number} = {line: -1, num: -1};
+    private matchedIterLocation: { line: number, num: number } = { line: -1, num: -1 };
```
Comment (veggiesaurus) [inline src/components/FileInfo/FileInfoComponent.tsx]: as above

**Review by veggiesaurus — COMMENTED** [review]
@zarda can you fix the minor whitespace issues and the merge conflicts in your other PR (https://github.com/CARTAvis/carta-backend/pull/739) so we can merge this in?

File: src/components/FileInfo/FileInfoComponent.tsx (hunk 7 above)
Comment (zarda) [reply src/components/FileInfo/FileInfoComponent.tsx]: Retrieved

**Review by veggiesaurus — APPROVED** [review]

**Comment by veggiesaurus** [comment]
@zarda I've made some changes to how you were handling the ranges, and fixed some spelling issues. @duidae can you have a look as well?

**Review by veggiesaurus — CHANGES_REQUESTED** [review]
@zarda a few other issues I picked up that need fixing:
- ~~We should be able to save sub-images of 2D images (no channel selection required)~~ Sorry, this was due to my changes, don't worry about that :see_no_evil: 
- I think the default output format should be the same as the input format. For example, if the file is FITS, the default output should be FITS as well. CASA should only be default if the file itself is a CASA file.

**Comment by zarda** [comment]
The spectral range could not be updated synchronously while we change the coordinate system or source image after 3bb3460

**Comment by veggiesaurus** [comment]
@zarda it previously worked because you were breaking MobX strict mode rules. I have adjusted this now, it should work. @kswang1029 please check

**Review by veggiesaurus — APPROVED** [review]

**Review by kswang1029 — APPROVED** [review]
feature-wise it is good.
